### PR TITLE
[perftools] Dump Timer Information in Seconds

### LINKED ISSUE #179: [perftools] Dump Timer Information in Seconds
Description
-------------

Currently, we dump timer information in cycles, which is not user-friendly. Instead it would be nice to report this statistics in seconds.

Caveats
-------------

- We get cycles information using the `rdtscp` instruction. Therefore, in order to present reliable information to the end user, by simply dividing the number of cycles by the operating frequency, we must ensure that the time stamp counter runs in invariant mode. Otherwise, we should come up with some approximation strategy.

### PR DESCRIPTION
**Pull request by ihchoi12**
## Description
* This PR fixes #179 

## Summary of Changes
* Approximately measure nanoseconds-per-cycle at initialization
* Print out duration in both cycles and nanoseconds

### CODE REVIEW

**Review by ppenna — APPROVED** [review]

File: src/rust/perftools/profiler/mod.rs [hunk 1]
```diff
@@ -302,12 +307,29 @@ impl Profiler {
...
         for root in self.roots.iter() {
-            root.borrow().write_recursive(out, total_duration, 0, max_depth)?;
+            root.borrow().write_recursive(out, total_duration, 0, max_depth, self.ns_per_cycle)?;
         }
 
         out.flush()
     }
 
+    fn measure_ns_per_cycle() -> f64 {
+        let start = SystemTime::now();
```
Comment (ppenna) [inline src/rust/perftools/profiler/mod.rs:317]: Could you please annotate the following variable declaration with the concerned type?

File: src/rust/perftools/profiler/mod.rs [hunk 2]
```diff
@@ -302,12 +307,29 @@ impl Profiler {
...
+        let (start_cycle, _): (u64, u32) = unsafe { x86::time::rdtscp() };
+        
+        test::black_box((0..10000).fold(0, |old, new| old ^ new)); // dummy calculations for measurement
+
+        let (end_cycle, _): (u64, u32) = unsafe { x86::time::rdtscp() };
+        let since_the_epoch = SystemTime::now()
```
Comment (ppenna) [inline src/rust/perftools/profiler/mod.rs:323]: Could you please annotate the following variable declaration with the concerned type?

File: src/rust/perftools/profiler/mod.rs [hunk 3]
```diff
@@ -302,12 +307,29 @@ impl Profiler {
...
+            .duration_since(start)
+            .expect("Time went backwards");
+        let in_ns = since_the_epoch.as_secs() * 1_000_000_000 +
```
Comment (ppenna) [inline src/rust/perftools/profiler/mod.rs:326]: Could you please annotate the following variable declaration with the concerned type?

**Review by ppenna — CHANGES_REQUESTED** [review]
Looks good.

Before I proceed with the merging, could you:
- Address the minor points that I highlighted

File: src/rust/perftools/profiler/mod.rs (hunk 3 above)
Comment (ihchoi12) [reply src/rust/perftools/profiler/mod.rs]: Done. Thanks for the review!

**Comment by ppenna** [comment]
This PR was merged in commit 48c450740bd31d254d0de2b0c11220e889a5079f.

**Pull request closed without merging**